### PR TITLE
Copy into partition by

### PR DESCRIPTION
**Pull request by samansmink**
This PR introduces a first version of the partitioned COPY operator. This PR builds on previous work from @lnkuiper adding the `PartitionedColumnData` (https://github.com/duckdb/duckdb/pull/4970) and the per thread output from @hannes (https://github.com/duckdb/duckdb/pull/5412)

To summarize, the Partitioned COPY:
- Supports both CSV and Parquet 
- Currently fully materializes data during partitioning (to be improved in follow-up pr's)
- Outputs 1 file per partition per thread (similar to https://github.com/duckdb/duckdb/pull/5412)

The partitioned write is used similarly to the PER_THREAD_OUTPUT feature:
```
COPY table TO '__TEST_DIR__/partitioned' (FORMAT PARQUET, PARTITION_BY (part_col_a, part_col_b));
```

this command will write files in this format, which is known as the hive partitioning scheme:

```
__TEST_DIR__/partitioned/part_col_a=<val>/part_col_b=<val>/data_<thread_number>.parquet
``` 

Partioned copy to S3 also works:
```
COPY table TO 's3://mah-bucket/partitioned' (FORMAT PARQUET, PARTITION_BY (part_col_a, part_col_b));
```

Finally, a check is performed for existing files/directories which is currently quite conservative (and on S3 will add a bit of latency). To disable this check and force writing, an ALLOW_OVERWRITE flag is added:
```
COPY table TO '__TEST_DIR__/partitioned' (FORMAT PARQUET, PARTITION_BY (part_col_a, part_col_b), ALLOW_OVERWRITE TRUE);
```
Note that this also works with the PER_THREAD_OUTPUT feature.

### Implementation
To support these features, a new class `HivePartitionedColumnData` is introduced, which implements the `PartitionedColumnData` interface from https://github.com/duckdb/duckdb/pull/4970. The main complexity here is that the `HivePartitionedColumnData` class needs to be able to discover new partitions in parallel. For the RadixPartitioning that was already implemented, the number of partitions is known in advance and does not change. Partition discovery is handled by all threads while writing tuples to their local `HivePartitionedColumnData`. To prevent expensive locking when synchronizing the partition discovery, each thread will have a local partition map to do quick lookups during partitioning, with a shared global state where new partitions are added. This means that only when adding new partitions to the global state a lock is required. Since this partitioned write is not expected to scale to super large amounts of partitions anyway, this should work well. Due to this shared state the partition indices between the thread local `HivePartitionedColumnData` objects will remain in sync.

### Benchmarks
Here's some rough benchmarks to give an indication of the performance overhead on a M1 macbook:

| COPY lineitem_sf1 to file.parquet w/ 8 threads | Avg   | Relative to fastest |
| ---------------------------------------------- | ----- | ------------------- |
| Regular copy                                   | 7.50  | 443.98%             |
| Disable preserve order (https://github.com/duckdb/duckdb/pull/5756)                         | 1.40  | 1.64%               |
| Threads (https://github.com/duckdb/duckdb/pull/5412)                                        | 1.38  | 0.00%               |
| Threads + disable order preserving             | 1.42  | 2.73%               |
| Hive Partitioned (4 partitions over 2 cols)    | 1.62  | 17.71%              |
| Hive Partitioned (28 partitions over 3 cols)   | 1.92  | 38.93%              |
| Hive Partitioned (84 paritions over 2 cols)    | 1.89  | 36.93%              |
| Hive Partitioned (432 paritions over 3 cols)   | 2.58  | 86.83%              |
| Hive Partitioned (1160 paritions over 5 cols)  | 5.67  | 310.78%             |
| Hive Partitioned (2526 partitions over 1 col)  | 11.15 | 708.10%             |

Note that performance for low amounts of partitions is very good. Higher amounts of partitions get pretty bad, but this is most likely due to the fact that at this partition count, the resulting files only contain very few tuples, leading to large IO overhead. I would expect that for larger files the relative overhead will be lower, but more benchmarking here is required.

### Transform partition columns before writing
Note that partition values are converted to strings. There are probably many edge cases where this won't work nicely. However, just using SQL in your copy statement you can transform the partition by columns however you want. For example to add a prefix to an existing column call part_col you could do:
```
COPY (SELECT * EXCLUDE (part_col), 'prefix-'::VARCHAR || part_col::VARCHAR as part_col FROM test) 
TO '__TEST_DIR__/partitioned' (FORMAT PARQUET, PARTITION_BY (part_col));
```

### Limitations
The partitioned write fully materializes, so it will quickly fill the available memory for large tables. This is not necessarily a blocker as the buffer manager can offload to disk. This will however mean that enough local disk space is required and for large files the data is: 
- stored in memory
- offloaded to disk
- read from disk
- written to the final partition file

This is not ideal, but will still be good enough for many use cases.

### Future work
The ideal partitioning COPY would produce 1 file per partition while providing mechanisms to limit memory usage and not require full materialization.

First step towards this goal is to make a streaming variant that can flush partitions as partition tuple limits (or possibly global operator limits) are reached. This would produce a single file per partition, per flush of that partition. 

The second step would be to not close files after flushing, allowing multiple flushes to a single file. With this we would achieve the desired behaviour where we can produce a single file per partition in a streaming fashion.

We have some nice idea's on implementing the above, so hopefully two more PR's coming up soon-ish :)

### CODE REVIEW

**Comment by samansmink** [comment]
> Perfect, now just add partitions column 🤡😜

@djouallah there ya go 😁

**Review by lnkuiper — APPROVED** [review]
The code looks great, and the performance numbers are awesome! Performance seems to degrade quite slowly, even with ~400 partitions. In the example with 2526 partitions, I think the performance degradation is not only due to I/O, but also due to writing the data to the in-memory partitions. With so few partitions, we are basically appending 1-2 tuples to each of the partitions at a time, which brings us down to tuple-at-a-time processing 😓 not much you can do about this, though. If you're interested, you could try sorting the data by the column that you're partitioning on. This should greatly speed up the partitioning code since more tuples are being written to a partition at a time.

Just one comment: Is it worth making a `.benchmark` file so that the CI checks that the performance does not degrade?

**Comment by tobilg** [comment]
Great work, this will be a killer feature for DuckDB! Do you have plans to support writing to S3 as well in the future?

**Comment by samansmink** [comment]
@lnkuiper thanks for the review, will add some benchmarks later today!

@tobilg wow somehow I completely forgot to test this.. Thanks for the comment 😅  This feature should just work™ with S3, but it may need some additional tweaks. Ill add the tests later today. An important use case for the partitioning feature is certainly to be able to repartition a dataset from and to S3 in a fully streaming fashion.

**Comment by tobilg** [comment]
> This feature should just work™ with S3, but it may need some additional tweaks. Ill add the tests later today. An important use case for the partitioning feature is certainly to be able to repartition a dataset from and to S3 in a fully streaming fashion

Thanks, that's great news! Will there be a limitation regarding the number of partitions?

**Comment by samansmink** [comment]
@tobilg There's for sure going to be a prohibitive overhead when partitioning with very large amounts of partitions, see also the comment by @lnkuiper and the benchmarks in this PR. However, with the optimizations we have planned (and perhaps the sorting optimization proposed by @lnkuiper) we should be able to support decently large amounts of partitions.

**Comment by tobilg** [comment]
> we should be able to support decently large amounts of partitions.

That'll be a great and very useful feature then! Other common ways on AWS to repartition parquet data in S3 have pretty tight limits, e.g. Athena CTAS queries can [only write to 100 partitions simulaneously](https://docs.aws.amazon.com/athena/latest/ug/ctas-considerations-limitations.html#ctas-considerations-limitations-partition-and-bucket-limits). One can work around this by using multiple passed, or pre-partitioning the data, but the DX is not great.

**Comment by lnkuiper** [comment]
@Mytherin This is a good idea. It is tricky to implement, however, because we are not scattering to a row layout but appending to an intermediate buffer, which is a `DataChunk`. Maybe we can discuss the options tomorrow.

**Comment by samansmink** [comment]
@lnkuiper @Mytherin I added two simple microbenchmarks which i've also added to the `.github/regression/micro.csv` however, since these write to files i'm not 100% this will work well as it may trigger false positives for regression

@tobilg Added the S3 tests. There were also some minor issues that are now resolved. So S3 seems to work fine!

File: src/common/hive_partitioning.cpp [hunk 1]
```diff
@@ -126,8 +131,120 @@ void HivePartitioning::ApplyFiltersToFileList(ClientContext &context, vector<str
...
+	for (auto it = global_state->partitions.begin() + local_partition_map.size(); it < global_state->partitions.end();
+	     it++) {
+		local_partition_map[(*it)->first] = (*it)->second;
+	}
+}
+
+idx_t HivePartitionedColumnData::RegisterNewPartition(HivePartitionKey key, PartitionedColumnDataAppendState &state) {
+	if (global_state) {
+		// We need to lock both the GlobalHivePartitionState and the allocators while adding a partition
+		unique_lock<mutex> lck_gstate(global_state->lock);
```
Comment (Mytherin) [inline src/common/hive_partitioning.cpp:226]: Do we need to hold two separate locks at the same time here? Using multiple locks is quite error-prone due to lock inversion.

File: src/common/hive_partitioning.cpp [hunk 2]
```diff
@@ -126,8 +131,120 @@ void HivePartitioning::ApplyFiltersToFileList(ClientContext &context, vector<str
...
+	D_ASSERT(allocators->allocators.size() == required_partitions);
+
+	for (idx_t i = current_partitions; i < required_partitions; i++) {
+		partitions.emplace_back(CreatePartitionCollection(i));
+		partitions[i]->InitializeAppend(*state.partition_append_states[i]);
+	}
+	D_ASSERT(partitions.size() == local_partition_map.size());
+}
+
+void HivePartitionedColumnData::SynchronizeLocalMap() {
```
Comment (Mytherin) [inline src/common/hive_partitioning.cpp:215]: It seems like these methods all require certain locks to be held - could we perhaps add them as parameters in a similar fashion to the `ClientContextLock`? That way the methods cannot be called without holding the lock.

File: src/execution/operator/persistent/physical_copy_to_file.cpp [hunk 3]
```diff
@@ -57,13 +74,69 @@ SinkResultType PhysicalCopyToFile::Sink(ExecutionContext &context, GlobalSinkSta
...
+
+static string CreateDirRecursive(ClientContext &context, const vector<idx_t> &cols, const vector<string> &names,
+                                 const vector<Value> &values, string path) {
+	CreateDir(context, path);
+
+	for (idx_t i = 0; i < cols.size(); i++) {
+		auto partition_col_name = names[cols[i]];
+		auto partition_value = values[i];
+		string p_dir = partition_col_name + "=" + partition_value.ToString();
+		path += "/" + p_dir;
```
Comment (Mytherin) [inline src/execution/operator/persistent/physical_copy_to_file.cpp:92]: Should this not get the path separator from the file system, or use `fs.JoinPath`?

File: src/execution/operator/persistent/physical_copy_to_file.cpp [hunk 4]
```diff
@@ -98,8 +186,9 @@ unique_ptr<LocalSinkState> PhysicalCopyToFile::GetLocalSinkState(ExecutionContex
 			this_file_offset = g.last_file_offset++;
 		}
 		auto &fs = FileSystem::GetFileSystem(context.client);
-		string output_path = fs.JoinPath(file_path, StringUtil::Format("out_%llu", this_file_offset));
-		if (fs.FileExists(output_path)) {
+		string output_path =
+		    fs.JoinPath(file_path, StringUtil::Format("out_%llu", this_file_offset) + "." + function.extension);
+		if (fs.FileExists(output_path) && !allow_overwrite) {
 			throw IOException("%s exists", output_path);
```
Comment (Mytherin) [inline src/execution/operator/persistent/physical_copy_to_file.cpp:192]: Perhaps mention the `allow_overwrite` option in the error here?

File: test/sql/copy/hive_filter_pushdown_bug.test [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+query II nosort q1
+explain SELECT * FROM parquet_scan('__TEST_DIR__/hive_pushdown_bug/*/*.parquet', HIVE_PARTITIONING=1) where a > b;
+----
+
+query II nosort q1
+explain SELECT * FROM parquet_scan('__TEST_DIR__/hive_pushdown_bug/*/*.parquet') where a > b;
+----
+
+query II
+explain SELECT * FROM parquet_scan('__TEST_DIR__/hive_pushdown_bug/*/*.parquet', HIVE_PARTITIONING=1) where a > c;
```
Comment (Mytherin) [inline test/sql/copy/hive_filter_pushdown_bug.test:25]: Unrelated to this PR but I wonder if we could auto-detect the `HIVE_PARTITIONING` by default from the set of globbed files, and make the `HIVE_PARTITIONING` parameter ternary (`false, true, auto` with default on `auto`).

File: third_party/libpg_query/src_backend_parser_scan.cpp [hunk 6]
```diff
@@ -1,5 +1,4 @@
-#line 2 "third_party/libpg_query/src_backend_parser_scan.cpp"
-#line 2 "third_party/libpg_query/scan.l"
+#line 1 "third_party/libpg_query/src_backend_parser_scan.cpp"
```
Comment (Mytherin) [inline third_party/libpg_query/src_backend_parser_scan.cpp:1]: It should not be necessary to re-generate the lexer unless you edit `scan.l` (which is rarely required). Could you revert the changes to this file?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Very exciting feature. Some comments from my side:

File: third_party/libpg_query/src_backend_parser_scan.cpp (hunk 6 above)
Comment (samansmink) [reply third_party/libpg_query/src_backend_parser_scan.cpp]: oops, reverted!

File: test/sql/copy/hive_filter_pushdown_bug.test (hunk 5 above)
Comment (samansmink) [reply test/sql/copy/hive_filter_pushdown_bug.test:25]: Good idea! Sounds like a nice first or second issue for one of our new devs

File: src/common/hive_partitioning.cpp (hunk 1 above)
Comment (samansmink) [reply src/common/hive_partitioning.cpp]: Good one, i took a better look at the locking here, and we definitely don't need to hold two locks, changed this to:

- acquire global state lock to sync global to local and vice versa
- release lock
- grow allocators (with allocator lock) to the required size for this thread. This is enough to guarantee there are always enough allocators available for each thread

File: src/common/hive_partitioning.cpp (hunk 2 above)
Comment (samansmink) [reply src/common/hive_partitioning.cpp:217]: on further inspection, 2/3 don't even need locking, and after fixing your previous comment, I moved the lock into the GrowAllocators method

File: src/common/bind_helpers.cpp [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+vector<bool> ParseColumnList(const vector<Value> &set, vector<string> &names, const string &loption) {
+	vector<bool> result;
+
+	if (set.empty()) {
+		throw BinderException("\"%s\" expects a column list or * as parameter", loption);
+	}
+	// list of options: parse the list
+	unordered_map<string, bool> option_map;
```
Comment (Mytherin) [inline src/common/bind_helpers.cpp:24]: `case_insensitive_map` perhaps?

**Review by Mytherin — COMMENTED** [review]
Thanks for the fixes! One minor comment, otherwise looks great:

File: src/common/bind_helpers.cpp (hunk 7 above)
Comment (samansmink) [reply src/common/bind_helpers.cpp]: ah good catch, added it + a test

**Review by Mytherin — APPROVED** [review]
Thanks for the updates - LGTM! Ready to merge after CI passes.

**Comment by samansmink** [comment]
@Mytherin 
there's 2 failures left:
- codecov fails with http error, can be restarted but it has passed in previous runs with little changes so i think its fine
- odbc failure on linux aarch64 is also http thing and unrelated, it also has succeeded in previous runs
